### Trying other models 

### Getting Started

As already we have our cleaned training data and cleaned test data so in this notebook we will just focus to imporve our model

In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv('cleaned_train.csv',index_col=False)
data.head()

Unnamed: 0  Kolkata_Average_Price  Kolkata_Ref_Price  \
0          30                 120.93              97.27   
1           2                  95.95              71.01   
2           3                  94.14              73.38   
3           4                  91.45              70.39   
4          17                 126.49              95.14   

   Bangalore_Average_Price  Bangalore_Ref_Price  Cochin_Average_Price  \
0                   112.08                92.36                105.19   
1                    87.66                71.01                 80.58   
2                    85.69                65.66                  0.00   
3                     0.00                64.99                 79.27   
4                   122.25                87.51                118.95   

   Cochin_Ref_Price  Darjeeling_Average_Price  Darjeeling_Ref_Price  \
0             87.63                      0.00                 72.88   
1             67.16                     82.25                 57.49   
2             65.57                     80.87                 54.59   
3             62.09                     80.76                 57.06   
4             86.36                     87.37                 59.05   

   Ernakulam_Average_Price  Ernakulam_Ref_Price  Siliguri_Average_Price  \
0                    69.64                67.21                   76.80   
1                    69.64                52.48                   71.66   
2                     0.00                53.43                   71.12   
3                    69.65                53.38                   72.30   
4                    85.14                56.17                   83.31   

   Siliguri_Ref_Price  Guwahati_Average_Price  Guwahati_Ref_Price    Average  \
0               68.90                   71.04               65.90  85.217692   
1               53.18                   69.51               48.04  69.830000   
2               52.07                   69.14               48.50  67.846667   
3               52.50                   69.39               50.33  67.196923   
4               55.35                   79.73               53.50  85.451429   

   SaleDate  SaleMonth  SaleYear  
0         8          1      2009  
1        17          1      2009  
2        24          1      2009  
3        31          1      2009  
4         5          2      2009

In [3]:
data.drop('Unnamed: 0',axis=1,inplace=True)

In [4]:
data.head()

Kolkata_Average_Price  Kolkata_Ref_Price  Bangalore_Average_Price  \
0                 120.93              97.27                   112.08   
1                  95.95              71.01                    87.66   
2                  94.14              73.38                    85.69   
3                  91.45              70.39                     0.00   
4                 126.49              95.14                   122.25   

   Bangalore_Ref_Price  Cochin_Average_Price  Cochin_Ref_Price  \
0                92.36                105.19             87.63   
1                71.01                 80.58             67.16   
2                65.66                  0.00             65.57   
3                64.99                 79.27             62.09   
4                87.51                118.95             86.36   

   Darjeeling_Average_Price  Darjeeling_Ref_Price  Ernakulam_Average_Price  \
0                      0.00                 72.88                    69.64   
1                     82.25                 57.49                    69.64   
2                     80.87                 54.59                     0.00   
3                     80.76                 57.06                    69.65   
4                     87.37                 59.05                    85.14   

   Ernakulam_Ref_Price  Siliguri_Average_Price  Siliguri_Ref_Price  \
0                67.21                   76.80               68.90   
1                52.48                   71.66               53.18   
2                53.43                   71.12               52.07   
3                53.38                   72.30               52.50   
4                56.17                   83.31               55.35   

   Guwahati_Average_Price  Guwahati_Ref_Price    Average  SaleDate  SaleMonth  \
0                   71.04               65.90  85.217692         8          1   
1                   69.51               48.04  69.830000        17          1   
2                   69.14               48.50  67.846667        24          1   
3                   69.39               50.33  67.196923        31          1   
4                   79.73               53.50  85.451429         5          2   

   SaleYear  
0      2009  
1      2009  
2      2009  
3      2009  
4      2009

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 544 entries, 0 to 543
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Kolkata_Average_Price     544 non-null    float64
 1   Kolkata_Ref_Price         544 non-null    float64
 2   Bangalore_Average_Price   544 non-null    float64
 3   Bangalore_Ref_Price       544 non-null    float64
 4   Cochin_Average_Price      544 non-null    float64
 5   Cochin_Ref_Price          544 non-null    float64
 6   Darjeeling_Average_Price  544 non-null    float64
 7   Darjeeling_Ref_Price      544 non-null    float64
 8   Ernakulam_Average_Price   544 non-null    float64
 9   Ernakulam_Ref_Price       544 non-null    float64
 10  Siliguri_Average_Price    544 non-null    float64
 11  Siliguri_Ref_Price        544 non-null    float64
 12  Guwahati_Average_Price    544 non-null    float64
 13  Guwahati_Ref_Price        544 non-null    float64
 14  Average   

In [6]:
data.isnull().sum()

Kolkata_Average_Price       0
Kolkata_Ref_Price           0
Bangalore_Average_Price     0
Bangalore_Ref_Price         0
Cochin_Average_Price        0
Cochin_Ref_Price            0
Darjeeling_Average_Price    0
Darjeeling_Ref_Price        0
Ernakulam_Average_Price     0
Ernakulam_Ref_Price         0
Siliguri_Average_Price      0
Siliguri_Ref_Price          0
Guwahati_Average_Price      0
Guwahati_Ref_Price          0
Average                     0
SaleDate                    0
SaleMonth                   0
SaleYear                    0
dtype: int64

### Modeling

In [7]:
from sklearn.model_selection import train_test_split, RandomizedSearchCV, GridSearchCV
from sklearn.linear_model import Ridge, Lasso, ElasticNet, SGDRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.ensemble import GradientBoostingRegressor, AdaBoostRegressor, BaggingRegressor
from sklearn.neighbors import KNeighborsRegressor

# Spliiting X and y
np.random.seed(42)

X = data.drop('Average',axis=1)
y = data['Average']

X_train, X_val, y_train, y_val = train_test_split(X,y,test_size=42)

len(X_train), len(X_val), len(y_train), len(y_val)

(502, 42, 502, 42)

In [8]:
models = {
    'Ridge': Ridge(),
    'Lasso' : Lasso(),
    'ElasticNet': ElasticNet(),
    'linear svr': SVR(kernel='linear'),
    'rbf svr':SVR(),
    'SGD': SGDRegressor(),
    'Gradient':GradientBoostingRegressor(),
    'Ada': AdaBoostRegressor(),
    'Bagging': BaggingRegressor(),
    'Kneighbor':KNeighborsRegressor()
}

In [9]:
#RMSE metric
def rmse(y_true,y_preds):
    return np.sqrt(mean_squared_error(y_true,y_preds))

In [10]:
def fit_score(model,X_train,X_valid,y_train,y_valid):
    for label,model in models.items():
        np.random.seed(42)
        
        #fitting the model
        model.fit(X_train,y_train)
        
        #predicting on validation 
        prediction_valid = model.predict(X_val)
        
        #MAE on validation set
        MAE_valid = mean_absolute_error(y_valid,prediction_valid)
        print(f'MAE_valid:{model} = {MAE_valid}')
        
        #RMSE on validation set
        RMSE_valid = rmse(y_valid,prediction_valid)
        print(f'RMSE_valid:{model} = {RMSE_valid}')
        

In [11]:
fit_score(
    model = models,
    X_train = X_train,
    X_valid = X_val,
    y_train = y_train,
    y_valid = y_val
)

MAE_valid:Ridge() = 3.9586695496060478
RMSE_valid:Ridge() = 5.269597885821289
MAE_valid:Lasso() = 3.7749062350200666
RMSE_valid:Lasso() = 5.22126241766144
MAE_valid:ElasticNet() = 3.7605382554738207
RMSE_valid:ElasticNet() = 5.2708849158891935
MAE_valid:SVR(kernel='linear') = 3.977902610137642
RMSE_valid:SVR(kernel='linear') = 5.808381697462686
MAE_valid:SVR() = 11.961593174517573
RMSE_valid:SVR() = 14.506794890004166
MAE_valid:SGDRegressor() = 3222891132077099.5
RMSE_valid:SGDRegressor() = 3226544700977289.0
MAE_valid:GradientBoostingRegressor() = 1.6328488951392155
RMSE_valid:GradientBoostingRegressor() = 3.2575735695089256
MAE_valid:AdaBoostRegressor() = 3.626379653250769
RMSE_valid:AdaBoostRegressor() = 5.591907634382114
MAE_valid:BaggingRegressor() = 1.9149766020000005
RMSE_valid:BaggingRegressor() = 3.835969092507393
MAE_valid:KNeighborsRegressor() = 4.030546905523808
RMSE_valid:KNeighborsRegressor() = 7.31205360751375


As we can see GradientBoostingRegressor and Bagging quiter perform well on default parameters so let's hypertune our these two models.

### Hypertuning our model

So let's tune our model by hand

In [12]:
#for bagging regressor
n_estimators = [10,20,30,40,50,60,70,80,90,100]
np.random.seed(42)
for i in n_estimators:
    bagging_model = BaggingRegressor(n_estimators=i)
    bagging_model.fit(X_train,y_train)
    bagging_preds = bagging_model.predict(X_val)
    bagging_score = rmse(y_val,bagging_preds)
    print(f'Number of estimators is {i} and score is {bagging_score}')

Number of estimators is 10 and score is 3.835969092507393
Number of estimators is 20 and score is 3.6983317661376116
Number of estimators is 30 and score is 3.588823136574166
Number of estimators is 40 and score is 3.7253650609547497
Number of estimators is 50 and score is 3.49085921031683
Number of estimators is 60 and score is 3.5201917161536933
Number of estimators is 70 and score is 3.6408884706790072
Number of estimators is 80 and score is 3.6636580551073203
Number of estimators is 90 and score is 3.694645183857085
Number of estimators is 100 and score is 3.6992817472462756


As we can see our **bagging_model** perform good on **n_estiamtors = 60** compare to b aseline model

In [13]:
#for Gradient Boosting regressor
n_estimators_gradient = [100,190,200,210,900,1000,1100]
np.random.seed(42)
for i in n_estimators_gradient:
    gradient_model = GradientBoostingRegressor(n_estimators=i)
    gradient_model.fit(X_train,y_train)
    gradient_preds = gradient_model.predict(X_val)
    gradient_score = rmse(y_val,gradient_preds)
    print(f'Number of estimators is {i} and score is {gradient_score}')

Number of estimators is 100 and score is 3.2575735695089256
Number of estimators is 190 and score is 3.212696559753233
Number of estimators is 200 and score is 3.294720356622678
Number of estimators is 210 and score is 3.2927086166958435
Number of estimators is 900 and score is 3.235873702416429
Number of estimators is 1000 and score is 3.1980825012559158
Number of estimators is 1100 and score is 3.237915860032344


Compare to bagging model our **gradient booting** model perform well when **number of estimator is 1000**

Let's fit our gradient boosting with these parameters

In [14]:
model_1 = GradientBoostingRegressor(n_estimators=1000,random_state=42)
model_1.fit(X_train,y_train)

GradientBoostingRegressor(n_estimators=1000, random_state=42)

In [15]:
model_1_preds = model_1.predict(X_val)

In [16]:
model_1_rmse = rmse(y_val,model_1_preds)
model_1_rmse

3.223349233002795

### Getting test data

In [17]:
test_data = pd.read_csv('cleaned_test.csv')

In [18]:
test_data.drop('Unnamed: 0',axis=1,inplace=True)

In [19]:
test_data

Kolkata_Average_Price  Kolkata_Ref_Price  Bangalore_Average_Price  \
0                  171.35             165.87                  154.080   
1                  171.35             173.12                  154.080   
2                  171.35             162.13                  156.140   
3                  171.35             162.13                  158.040   
4                  171.35             162.13                  157.550   
5                  179.47             162.13                  154.080   
6                  171.35             162.13                  163.940   
7                  182.14             162.13                  154.080   
8                  183.59             162.13                  154.080   
9                  171.35             169.98                  154.080   
10                 174.81             162.13                  154.080   
11                 171.35             162.13                  154.080   
12                 171.35             161.30                  154.080   
13                 171.35             162.13                  149.500   
14                 171.35             162.13                  154.080   
15                 162.83             160.65                  154.080   
16                 171.35             160.79                  154.080   
17                 171.35             164.21                  151.215   
18                 171.35             162.13                  154.080   
19                 171.35             162.13                  154.080   
20                 171.35             162.13                  154.080   
21                 155.60             162.13                  154.080   
22                 171.35             162.13                  154.080   
23                 171.35             162.13                  141.720   
24                 171.35             162.13                  154.080   
25                 171.35             159.12                  154.080   
26                 153.19             162.13                  154.080   
27                 171.35             162.13                  154.080   
28                 152.94             162.13                  134.250   

    Bangalore_Ref_Price  Cochin_Average_Price  Cochin_Ref_Price  \
0                160.82                132.70            142.63   
1                149.95                132.70            142.63   
2                149.95                132.70            150.39   
3                149.95                132.70            146.68   
4                163.85                132.70            144.87   
5                149.95                132.70            142.63   
6                149.95                132.70            142.63   
7                149.95                132.70            142.63   
8                160.61                138.75            139.98   
9                149.95                132.70            142.63   
10               149.95                132.70            142.63   
11               149.95                132.70            142.63   
12               149.95                132.70            140.40   
13               150.93                132.25            142.63   
14               149.95                132.70            142.63   
15               147.19                129.09            142.63   
16               149.95                130.74            142.63   
17               148.63                132.70            142.63   
18               149.95                132.70            142.63   
19               149.95                136.59            141.61   
20               149.95                132.70            142.63   
21               149.95                132.30            138.11   
22               149.95                132.70            142.63   
23               147.28                137.61            146.67   
24               149.95                132.70            142.63   
25               149.95                132.70            142.63   
26  

In [20]:
test_preds = model_1.predict(test_data)

In [21]:
submission_df = pd.DataFrame()
submission_df['Average'] = test_preds

In [22]:
submission_df.to_csv('my_second_submission_file.csv', index=False)

### Hypertuning our Gradient Boosting Regressor Model (Experimenting)

For this we going to make our **n_estimators = 1000** and further experimenting with other parameters **by hand**

In [23]:
#hypertuning learning_rate
learning_rate = [0.01,0.1,1]
np.random.seed(42)
for i in learning_rate:
    gb_model = GradientBoostingRegressor(n_estimators=1000,learning_rate=i)
    gb_model.fit(X_train,y_train)
    gb_preds = gb_model.predict(X_val)
    gb_score = rmse(y_val,gb_preds)
    print(f'Learning rate is {i} and score is {gb_score}')

Learning rate is 0.01 and score is 3.302683280318492
Learning rate is 0.1 and score is 3.208013872335586
Learning rate is 1 and score is 5.941931646537374


Our previous hypertuned model got **3.223** **RMSE** and this new hypertuned model got around **3.20 RMSE**


So let's further experimenting with our model with this parameters

let's fit our model with n_estimators = 1000 and learning_rate=0.1

In [24]:
#hypertuning max_depth
max_depth = [2,3,4,5,6,7,8]
np.random.seed(35)
for i in max_depth:
    gbm_model = GradientBoostingRegressor(n_estimators=1000,learning_rate=0.1,max_depth=i)
    gbm_model.fit(X_train,y_train)
    gbm_preds = gbm_model.predict(X_val)
    gbm_score = rmse(y_val,gbm_preds)
    print(f'Max depth is {i} and score is {gbm_score}')

Max depth is 2 and score is 2.668125277956219
Max depth is 3 and score is 3.259394321632904
Max depth is 4 and score is 3.627449229359672
Max depth is 5 and score is 3.765008213639044
Max depth is 6 and score is 3.723372439604142
Max depth is 7 and score is 3.864597483181562
Max depth is 8 and score is 4.030259003613861


Woah that's pretty score jumping from **3.20** to **2.66** with **max_depth = 2**

### Let's fit our model with these parameters


In [25]:
np.random.seed(35)
model_2 = GradientBoostingRegressor(n_estimators=1000,learning_rate=0.1,max_depth=2)
model_2.fit(X_train,y_train)
preds_2 = model_2.predict(X_val)
rmse_2 = rmse(y_val,preds_2)
rmse_2

2.668125277956219

### Getting Test Predictions with this new hypertuned model

In [26]:
test_preds_2 = model_2.predict(test_data)

In [27]:
submission_df_2 = pd.DataFrame()
submission_df_2['Average'] = test_preds_2


In [28]:
submission_df_2

Average
0   119.650813
1   117.596354
2   119.787465
3   119.440412
4   119.181463
5   116.253570
6   118.132431
7   118.951551
8   121.138722
9   119.691814
10  118.574552
11  118.584194
12  120.589456
13  116.929737
14  119.662014
15  118.311419
16  118.407216
17  118.836698
18  119.520500
19  119.335444
20  119.809804
21  117.197959
22  119.459662
23  119.515891
24  121.255701
25  119.447630
26  117.240679
27  119.645454
28  117.472369

In [29]:
submission_df.to_csv('my_third_submission_file.csv', index=False)


### Experimneting

In [30]:
loss = ['ls','lad','huber']

np.random.seed(35)
for i in loss:
    gbml_model = GradientBoostingRegressor(n_estimators=1000,learning_rate=0.1,max_depth=2,loss=i)
    gbml_model.fit(X_train,y_train)
    gbml_preds = gbml_model.predict(X_val)
    gbml_score = rmse(y_val,gbml_preds)
    print(f'Loss is {i} and score is {gbml_score}')

Loss is ls and score is 2.668125277956219
Loss is lad and score is 3.082174406739605
Loss is huber and score is 2.9721927633941716


In [33]:
min_samples_split = [2,3,4,5,6,7,8]
np.random.seed(35)
for i in min_samples_split:
    gbmls_model = GradientBoostingRegressor(n_estimators=1000,learning_rate=0.1,max_depth=2,loss='ls',min_samples_split=i)
    gbmls_model.fit(X_train,y_train)
    gbmls_preds = gbmls_model.predict(X_val)
    gbmls_score = rmse(y_val,gbmls_preds)
    print(f'min_samples_split is {i} and score is {gbmls_score}')

min_samples_split is 2 and score is 2.668125277956219
min_samples_split is 3 and score is 2.6789303055672367
min_samples_split is 4 and score is 2.677370718906436
min_samples_split is 5 and score is 2.6754255686531176
min_samples_split is 6 and score is 2.6896556048200995
min_samples_split is 7 and score is 2.6901343854515303
min_samples_split is 8 and score is 2.61298079738006


Let's take **min_samples_split = 8**

In [38]:
min_samples_leaf = [2,3,4,5,6,7,8]
np.random.seed(35)
for i in min_samples_leaf:
    gbmlsl_model = GradientBoostingRegressor(n_estimators=1000,learning_rate=0.1,max_depth=2,loss='ls',min_samples_split=8,min_samples_leaf=i)
    gbmlsl_model.fit(X_train,y_train)
    gbmlsl_preds = gbmlsl_model.predict(X_val)
    gbmlsl_score = rmse(y_val,gbmlsl_preds)
    print(f'min_samples_leaf is {i} and score is {gbmlsl_score}')

min_samples_leaf is 2 and score is 2.626050909393405
min_samples_leaf is 3 and score is 2.5918886508509584
min_samples_leaf is 4 and score is 2.6635635017805583
min_samples_leaf is 5 and score is 2.575415321606267
min_samples_leaf is 6 and score is 2.608473562071202
min_samples_leaf is 7 and score is 2.5946151002503117
min_samples_leaf is 8 and score is 2.726797399842577


In [39]:
max_features = ['auto', 'sqrt','log2']
np.random.seed(35)
for i in max_features:
    gbmlslm_model = GradientBoostingRegressor(
        n_estimators=1000,
        learning_rate=0.1,
        max_depth=2,
        loss='ls',
        min_samples_split=8,
        min_samples_leaf=5,
        max_features=i
    )
    gbmlslm_model.fit(X_train,y_train)
    gbmlslm_preds = gbmlslm_model.predict(X_val)
    gbmlslm_score = rmse(y_val,gbmlslm_preds)
    print(f'max feature is {i} and score is {gbmlslm_score}')

max feature is auto and score is 2.5754817045683276
max feature is sqrt and score is 2.643352312245121
max feature is log2 and score is 2.8207821659451295


After tuning our hypermeters by hand, let's fit our model with our best parameters

In [40]:
final_model = GradientBoostingRegressor(
        n_estimators=1000,
        learning_rate=0.1,
        max_depth=2,
        loss='ls',
        min_samples_split=8,
        min_samples_leaf=5,
        max_features='auto'
    )

final_model.fit(X_train,y_train)

GradientBoostingRegressor(max_depth=2, max_features='auto', min_samples_leaf=5,
                          min_samples_split=8, n_estimators=1000)

In [41]:
final_preds = final_model.predict(X_val)

In [42]:
final_rmse = rmse(y_val,final_preds)

In [43]:
final_rmse

2.575415321606267

### Predicting on test data

In [44]:
final_test_preds = final_model.predict(test_data)

In [45]:
final_test_preds

array([119.73947829, 117.48573241, 119.44402655, 119.11080552,
       119.2731768 , 116.0457075 , 118.1247529 , 118.87627044,
       120.96804104, 119.41442413, 117.98776839, 118.88548031,
       120.5458375 , 116.4826444 , 119.54856164, 118.70464909,
       118.53552358, 119.10974376, 119.54042385, 119.26480449,
       119.77451119, 117.1101971 , 119.53167251, 119.61042501,
       121.35620577, 119.1177061 , 117.38206105, 119.69593011,
       117.77813124])

In [46]:
final_submission = pd.DataFrame()

In [47]:
final_submission['Average'] = final_test_preds

In [48]:
final_submission.head()

Average
0  119.739478
1  117.485732
2  119.444027
3  119.110806
4  119.273177

In [50]:
final_submission.to_csv('my_fourth_submission.csv',index=False)